### DISCUSSION TITLE
How to automatically add the traceId to the HttpServletRequest

### DISCUSSION THREAD

**Discussion by mitht** [open-telemetry/opentelemetry-java-instrumentation#12295] [Q&A] [answered]
I am using Java zero-code instrumentation.  Is it possible to automatically add the traceId to the HttpServletRequest by modifying the source code or configuration of opentelemetry-java-instrumentation?

java -javaagent:path/to/opentelemetry-javaagent.jar -Dotel.service.name=your-service-name -jar myapp.jar

**Comment by laurit** [comment]
Sure, anything is possible when you modify the source code of the agent. Though you could also do it in your application. Could you elaborate what exactly you are attempting to do?

**Reply by mitht** [reply, marked as answer]
By using a non-intrusive approach to pass the traceId to the business code, it is possible to use Arthas's watch method to listen to the controller calls and obtain the traceId from the request headers.

versions:
spring-boot 2.6.6 
opentelemetry-javaagent-1.33.1.jar

In the io.opentelemetry.javaagent.instrumentation.servlet.BaseServletHelper class, the following lines are already set:

```java
accessor.setRequestAttribute(request, "trace_id", spanContext.getTraceId());
accessor.setRequestAttribute(request, "span_id", spanContext.getSpanId());

```
<img width="1514" alt="image" src="https://github.com/user-attachments/assets/fa1d7dcb-a479-4a51-a528-a7569c9a57c9">


However, in my business code, when I try to get the result using

```java
((ServletRequestAttributes) RequestContextHolder.getRequestAttributes()).getRequest().getAttribute("trace_id");

```

the result is null.

<img width="966" alt="image" src="https://github.com/user-attachments/assets/0fce5029-2388-4976-ae2e-681aea9e89f4">

Should the following code be added again in the Servlet2Advice class?
```java
io.opentelemetry.javaagent.instrumentation.servlet.v2_2.Servlet2Advice
SpanContext spanContext = Span.fromContext(context).getSpanContext();
String traceId = spanContext.getTraceId();
httpServletRequest.setAttribute("traceId", traceId);  
```
In the opentelemetry-java-instrumentation source code, in which instrumentation should the request header be set? In Servlet, HttpServletRequest does not support setHeader()
